File: gedcom/src/nu/mine/mosher/gedcom/GedcomLine.java
package nu.mine.mosher.gedcom;

//import nu.mine.mosher.util.Traversable;

public class GedcomLine //implements Traversable
{
	private final int level;
	private final String id;
	private final String tag;
	private String value;
	private final String pointer;

	public GedcomLine(int level, String id, String tag, String value)
	{
		this.level = level;
		this.id = getPointer(id);
		this.tag = tag;
		String v = getPointer(value);
		if (v.length() > 0)
		{
			this.pointer = v;
			this.value = "";
		}
		else
		{
			this.value = replaceAts(value);
			this.pointer = "";
		}
	}

	private static String getPointer(String s)
	{
		if (!s.startsWith("@") || !s.endsWith("@") || s.length() < 3)
		{
			return "";
		}

		String pointer = s.substring(1,s.length()-1);
		if (pointer.indexOf('@') >= 0)
		{
			return "";
		}
		return pointer;
	}

	private static String replaceAts(String s)
	{
		return s.replaceAll("@@","@");
	}

	@Override
	public String toString()
	{
		StringBuffer sb = new StringBuffer(256);

		sb.append(this.level);
		sb.append(",");
		if (hasID())
		{
			sb.append("id=");
			sb.append(this.id);
			sb.append(",");
		}
		sb.append("tag=");
		sb.append(this.tag);
		sb.append(",");
		if (isPointer())
		{
			sb.append("pointer=");
			sb.append(this.pointer);
		}
		else
		{
			sb.append("value=\"");
			sb.append(this.value);
			sb.append("\"");
		}

		return sb.toString();
	}

	public boolean hasID()
	{
		return this.id.length() > 0;
	}

	public boolean isPointer()
	{
		return this.pointer.length() > 0;
	}

    /**
     * @return
     */
    public String getId()
    {
        return this.id;
    }

    /**
     * @return
     */
    public int getLevel()
    {
        return this.level;
    }

    /**
     * @return
     */
    public String getPointer()
    {
        return this.pointer;
    }

    /**
     * @return
     */
    public String getTag()
    {
        return this.tag;
    }

    /**
     * @return
     */
    public String getValue()
    {
        return this.value;
    }

    /**
     * @param string
     */
    public void contValue(String string)
    {
    	this.value += "\n";
    	this.value += string;
    }

    /**
     * @param string
     */
    public void concValue(String string)
    {
		this.value += string;
    }
/*
    public void enter()
    {
    	System.out.print("<");
		System.out.print(this.tag);

		if (hasID())
		{
			System.out.print(" id=\"");
			System.out.print(this.id);
			System.out.print("\"");
		}

		if (isPointer())
		{
			System.out.print(" xref=\"");
			System.out.print(this.pointer);
			System.out.print("\"");
		}

		System.out.print(">");

		if (this.value.length() > 0)
		{
			System.out.print("<![CDATA[");
			System.out.print(this.value);
			System.out.print("]]>");
		}
		System.out.println();
    }

    public void leave()
    {
		System.out.print("</");
		System.out.print(this.tag);
		System.out.println(">");
    }
*/
}

File: genealogy/grojs/GedcomLine.js
/**
 * @fileoverview
 * Defines the {@link GedcomLine} class.
 */

/**
 * @class Represents one line of a GEDCOM file.
 * @requires Util
 * 
 * @constructor
 * @param {Number} level the level of this {@link GedcomLine}.
 * @param {String} gid the ID of this {@link GedcomLine}. Optional.
 * @param {String} tag the tag of this {@link GedcomLine}.
 * @param {String} val the value of this {@link GedcomLine}. Optional
 * @return new {@link GedcomLine}
 * @type GedcomLine
 */
function GedcomLine(level, gid, tag, val) {
	Util.verifyType(this,"GedcomLine");

	/**
	 * Level number. Greater than or equal to zero.
	 * @private
	 * @type Number
	 */
	this.level = parseInt(level,10);

	/**
	 * ID of this line, if it has one, otherwise empty string.
	 * @private
	 * @type String
	 */
	this.gid = GedcomLine.asPointer(gid);

	/**
	 * GEDCOM tag of this line (not validated).
	 * @private
	 * @type String
	 */
	this.tag = Util.safeStr(tag).toUpperCase();

	/**
	 * ID of other line if value is a pointer, otherwise empty string.
	 * @private
	 * @type String
	 */
	this.pointer = GedcomLine.asPointer(val);

	if (this.pointer) {
		/**
		 * Value of this line, or empty string.
		 * @private
		 * @type String
		 */
		this.val = Util.safeStr();
	} else {
		this.val = GedcomLine.replaceAts(Util.safeStr(val));
	}
}

/**
 * Gets the level number of this {@link GedcomLine}.
 * @return level
 * @type Number
 */
GedcomLine.prototype.getLevel = function() {
	return this.level;
};

/**
 * Gets the ID of this {@link GedcomLine}, or empty string if this
 * line doesn't have an ID.
 * @return ID
 * @type String
 */
GedcomLine.prototype.getID = function() {
	return this.gid;
};

/**
 * Gets the GECOM tag of this {@link GedcomLine}.
 * @return tag
 * @type String
 */
GedcomLine.prototype.getTag = function() {
	return this.tag;
};

/**
 * Gets the value of this {@link GedcomLine}. May be empty.
 * @return value
 * @type String
 */
GedcomLine.prototype.getVal = function() {
	return this.val;
};

/**
 * If the value of this {@link GedcomLine} is just a pointer to
 * another line, then this method gets the ID of that other line.
 * Otherwise, returns an empty string.
 * @return pointer
 * @type String
 */
GedcomLine.prototype.getPointer = function() {
	return this.pointer;
};

/**
 * Returns <code>true</code> if this {@link GedcomLine} has an ID.
 * @return if this {@link GedcomLine} has an ID
 * @type Boolean
 */
GedcomLine.prototype.hasID = function() {
	return !!this.getID();
};

/**
 * Returns <code>true</code> if this {@link GedcomLine}'s value
 * is just a pointer to another line.
 * @return if this {@link GedcomLine} is a pointer
 * @type Boolean
 */
GedcomLine.prototype.isPointer = function() {
	return !!this.getPointer();
};

/**
 * Joins a CONC or CONT {@link GedcomLine} to this {@link GedcomLine}.
 * If the given {@link GedcomLine} is not a CONC or CONT line, then do nothing.
 * @param {GedcomLine} c the CONC or CONT line
 */
GedcomLine.prototype.concat = function(c) {
	switch (c.getTag()) {
		case "CONC": this.val += c.getVal(); break;
		case "CONT": this.val += "\n"+c.getVal(); break;
	}
};

/**
 * Creates a {@link GedcomLine} from parsing the given GEDCOM line string.
 * @param {String} s the GEDCOM line to be parsed.
 * @returns new {@link GedcomLine} from parsing s
 * @type GedcomLine
 */
GedcomLine.parse = function(s) {
	var r = /^\s*(\d+)\s+(?:(@[^@]+@)\s+)?(\S+)(?:\s(.*))?$/.exec(s);
	if (r == null) {
		throw new Error("Gedcom line has invalid syntax: "+s);
	}
	return new GedcomLine(r[1],r[2],r[3],r[4]);
};

/**
 * Parses the given string to see if it is a GEDCOM pointer (that is, if
 * it starts and ends with an at sign and does not contain any other at signs).
 * If so, returns the value inside the at signs (the ID). If not, returns the
 * empty string.
 * @private
 * @param {String} s the value to check for being a GEDCOM pointer
 * @return the pointed to ID, or empty string
 * @type String 
 */
GedcomLine.asPointer = function(s) {
	var r;
	r = /^@([^@]+)@$/.exec(s);
	if (r == null) {
		return "";
	}
	return r[1];
};

/**
 * Replaces double at signs with single at signs in the given string.
 * @private
 * @param {String} s string containing double at-signs
 * @return string s with double at-signs replaced with single at-signs
 * @type String
 */
GedcomLine.replaceAts = function(s) {
	if (!s || !s.replace) {
		return s;
	}
	return s.replace(/@@/g,"@");
};
